---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# prime-seq


This repository contains scripts used for the analysis performed in our manuscript
  
**prime-seq, a simple, inexpensive, and powerful bulk RNA-sequencing protocol for detecting differential expression**

Aleksandar Janjic, Lucas E. Wange, Johannes W. Bagnoli, Johanna Geuder, Phong Nguyen, Daniel Richter, Beate Vieth, Christoph Ziegenhain, Binje Vick, Ines Hellmann, Wolfgang Enard

For the full prime-seq protocol please visit [protocols.io](https://www.protocols.io/view/prime-seq-s9veh66).

# <img src="Fig1.png" align="center" width="1000" />

Study Overview: prime-seq is a simple RNA-seq workflow that goes from lysate to sequencing library in no time. We benchmarked it's performance against the MAQC-III study using power analysis and showed that it captures know biological differences in a differentiation experiment.

The Data necessary to reproduce this analysis can be found at ArrayExpress:

| Accession    | Dataset              |
|--------------|----------------------|
| E-MTAB-10140 | Beads_Columns_tissue |
| E-MTAB-10138 | Beads_Columns_PBMC   |
| E-MTAB-10142 | Beads_Columns_HEK    |
| E-MTAB-10141 | gDNA_priming         |
| E-MTAB-10139 | UHRR                 |
| E-MTAB-10133 | iPSC                 |
| E-MTAB-10175 | AML                  |


## Preprocessing
All RNA-seq data was adapter trimmed with [cutadapt] (https://cutadapt.readthedocs.io/en/stable/) and preprocessed with [zUMIs](https://github.com/sdparekh/zUMIs/) (Parekh et al., 2017).

## 1. Figure: RNA isolation



## 2. Figure: power simulations
## 3. Figure: proof of concept, AML and iPSC to NPC
## 4. Figure: Budget vs. Power
## 5. Supplementary Figures
## 6. Molecular Workflow of prime-seq

# <img src="Workflow.png" align="center" width="1000" />


## `R` Session Info

```{r}
sessionInfo()
```