---
title: "prime-seq - gDNA Priming"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

In this experiment we tried to estimate the rate of gDNA priming in prime-seq. To generate an easily measurable readout we mixed gDNA from one species and RNA from another species in known ratios. This way we can deduce the rate of misspriming from the mapping to Species 1 or 2. The detailed experimental design can be found here: https://benchling.com/ajanjic/f/lib_hfKiJW6M-bulk-seq/etr_T0I2sva8-experiment-rna-vs-gdna-priming/edit

This Experiment was conducted with the E3V7 primer testset. As those are only 8 different barcodes we dual indexed the libraries. the Three replicates were run as individual zUMIs runs and need to be combined.

```{r load libraries and info file, message=F,warning=F}
library(tidyverse)
library(ggsci)
library(ggbeeswarm)
library(cowplot)
## theme

theme_pub <- theme_bw() + theme(
                                     plot.title = element_text(hjust = 0.5, size=18, face="bold"),
                                     axis.text = element_text(colour="black", size=14), 
                                     axis.title=element_text(size=16,face="bold"), 
                                     legend.text=element_text(size=14),
                                     legend.position="right",
                                     axis.line.x = element_line(colour = "black"), 
                                     axis.line.y = element_line(colour = "black"),
                                     strip.background=element_blank(), 
                                     strip.text=element_text(size=16))  


theme_set(theme_pub)

#### Colours
mapping_cols<- c("dodgerblue4","#086483","#F08C4B","#E5DFCC","#abd1f2","#eeccec","orangered1","grey33","dodgerblue4")
names(mapping_cols)<-c("Exon","Intron","Ambiguity","Intergenic","Intergenic:Human","Intergenic:Mouse","Unassigned","Unmapped","Assigned")

Contamination_cols<-c("#286687",
                      "#a36791",
                      "#e664d8",
                      "#d641aa",
                      "#a1e527",
                      "#802866")
names(Contamination_cols)<-c("No DNA Contamination",
                             "Low DNA Contamination",
                             "High DNA Contamination",
                             "Very High DNA Contamination",
                             "DNAse Treated",
                             "No RNA")

mouseman_cols<-c("#d641aa",
                 "#802866",
                 "#319eda",
                 "#286687",
                 "#abd1f2",
                 "#eeccec")
names(mouseman_cols)<-c("Intron:Mouse",
                        "Exon:Mouse",
                        "Intron:Human",
                        "Exon:Human",
                        "Intergenic:Human",
                        "Intergenic:Mouse")


fig_path<-"/data/share/htp/prime-seq_Paper/Fig_gdna_priming/"
```

```{r}
sample_info<-read.csv("/data/share/htp/prime-seq_Paper/Fig_gdna_priming/gDNA_priming_info.csv",stringsAsFactors = F)
## make unique cellbcs
sample_info$XC2<-paste0(sample_info$XC,"_",sample_info$Replicate)
sample_info<-sample_info %>% 
  mutate(Condition=case_when(Condition=="2RNA_0DNA"~"No DNA Contamination",
              Condition=="1.5RNA_0.5DNA"~"Low DNA Contamination",
              Condition=="2RNA_2DNA"~"High DNA Contamination",
              Condition=="0.5RNA_1.5DNA"~"Very High DNA Contamination",
              Condition=="2RNA_2DNA_DNAse"~"DNAse Treated",
              Condition=="0RNA_2DNA"~"No RNA")) %>% 
  mutate(Condition=factor(Condition,levels = c("No DNA Contamination","Low DNA Contamination","High DNA Contamination","Very High DNA Contamination","DNAse Treated","No RNA")))
```


## 1. Check mapping in bcstats

```{r}
## make list of dfs
path <- "/data/share/htp/prime-seq_Paper/Fig_gdna_priming/zUMIs"

dirs<-list.dirs(path, recursive = F)


feature_list<- list()

for (i in 1:length(dirs)){
  tmp<- strsplit(dirs[i], split = "/")[[1]][8]
  feature_list[[i]]<- read.table(paste0(dirs[i],
                                "/zUMIs_output/stats/Bulk_opt_gDNA_priming_",
                                tmp,
                                ".readspercell.txt"),
                                 sep= "\t",
                                 header=T)
  feature_list[[i]]$XC2<-paste0(feature_list[[i]]$RG,"_",strsplit(tmp,split = "rep")[[1]][2]) ## to make barcodes unique
  names(feature_list)[i]<-tmp
}


feature_df<-do.call("rbind", feature_list)

feature_df<-left_join(feature_df,sample_info)

feature_df$Replicate<-str_split_fixed(feature_df$XC2,pattern = "_",n=2)[,2]

feature_df<-feature_df %>% 
  filter(type!="User" & RG!="bad") %>% 
  mutate(type=factor(type,
                     levels = c("Ambiguity","Unmapped","Intergenic","Intron","Exon"))
                    ) 


p.map<-ggplot()+
  geom_bar(data=feature_df,aes(x=Condition,y=N,fill=type),stat="identity",
           position =position_fill())+
  coord_flip()+
  scale_fill_manual(values=mapping_cols)+
  theme(strip.text = element_text(size=6))+
  ylab("Mapped Reads")+
  xlab("")+
  theme(legend.position = "bottom", legend.title = element_blank(),legend.text = element_text(size=8))
p.map
```



```{r}
#Code from JWB

## make list of dfs

dirs<-list.dirs(path, recursive = F)


dge_list<- list()


for (i in 1:length(dirs)){
  tmp<- strsplit(dirs[i], split = "/")[[1]][8]
  dge_list[[i]]<- readRDS(paste0(dirs[i], "/zUMIs_output/expression/Bulk_opt_gDNA_priming_", tmp, ".dgecounts.rds"))
  names(dge_list)[i]<-tmp
}

### UMI counts
exon_list<- list()
for (i in 1:length(dge_list)){
  exon_list[[i]]<- as.matrix(dge_list[[i]]$umicount$exon$downsampling$downsampled_10000)
  colnames(exon_list[[i]])<-paste0(colnames(exon_list[[i]]),"_",i)
  names(exon_list)[i]<-names(dge_list)[i]
}

intron_list<- list()
for (i in 1:length(dge_list)){
  intron_list[[i]]<- as.matrix(dge_list[[i]]$umicount$intron$downsampling$downsampled_10000)
  colnames(intron_list[[i]])<-colnames(intron_list[[i]])<-paste0(colnames(intron_list[[i]]),"_",i)
  names(intron_list)[i]<-names(dge_list)[i]
}

## human only
ENSG_list_ex<- list()
for (i in 1:length(exon_list)){
  ENSG_list_ex[[i]]<- exon_list[[i]][grep("ENSG",row.names(exon_list[[i]])),]
  names(ENSG_list_ex)[i]<-names(exon_list)[i]
}

ENSG_list_in<- list()
for (i in 1:length(intron_list)){
  ENSG_list_in[[i]]<- intron_list[[i]][grep("ENSG",row.names(intron_list[[i]])),]
  names(ENSG_list_in)[i]<-names(intron_list)[i]
}

## mouse only
ENSMUS_list_ex<- list()
for (i in 1:length(exon_list)){
  ENSMUS_list_ex[[i]]<- exon_list[[i]][grep("ENSMUS",row.names(exon_list[[i]])),]
  names(ENSMUS_list_ex)[i]<-names(exon_list)[i]
}

ENSMUS_list_in<- list()
for (i in 1:length(intron_list)){
  ENSMUS_list_in[[i]]<- intron_list[[i]][grep("ENSMUS",row.names(intron_list[[i]])),]
  names(ENSMUS_list_in)[i]<-names(intron_list)[i]
}


df_list<-list()

for (i in 1:length(ENSG_list_ex)){
  df_list[[i]]<- data.frame(XC2=colnames(ENSG_list_ex[[i]]),
                            nUMI_HS_ex=colSums(ENSG_list_ex[[i]]),
                            nGene_HS_ex=colSums(ENSG_list_ex[[i]]>1),
                            nUMI_MM_ex=colSums(ENSMUS_list_ex[[i]]),
                            nGene_MM_ex=colSums(ENSMUS_list_ex[[i]]>1), 
                            nUMI_HS_in=colSums(ENSG_list_in[[i]]),
                            nGene_HS_in=colSums(ENSG_list_in[[i]]>1),
                            nUMI_MM_in=colSums(ENSMUS_list_in[[i]]),
                            nGene_MM_in=colSums(ENSMUS_list_in[[i]]>1))
}

df<- dplyr::bind_rows(df_list)


df2<- dplyr::left_join(df, sample_info)


```

```{r}

df2<-df2 %>% 
  group_by(Condition) %>% 
  mutate(contamination_umi=(nUMI_MM_ex+nUMI_MM_in)/(nUMI_MM_ex+nUMI_HS_ex+nUMI_MM_in+nUMI_MM_in)*100,
         contamination_gene=(nGene_MM_ex+nGene_MM_in)/(nGene_MM_ex+nGene_HS_ex+nGene_MM_in+nGene_MM_in)*100,
         mean_contamination_umi=mean(contamination_umi),
         mean_contamination_gene=mean(contamination_gene))


p.frac<-ggplot(data=df2,aes(y=contamination_umi,x=Condition,colour=Condition))+
  geom_quasirandom(width = 0.2)+
  geom_errorbar(aes(x=Condition,ymin=mean_contamination_umi,ymax=mean_contamination_umi,group=Condition,width=0.5))+
  ylab("% Contaminating UMIs")+
  xlab("")+
  scale_colour_manual(values=Contamination_cols)+
  coord_flip()+
  theme(legend.position = "none",
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())


p.frac

p.frac.g<-ggplot(data=df2,aes(y=contamination_gene,x=Condition,colour=Condition,shape=))+
  geom_quasirandom(width = 0.2)+
  geom_errorbar(aes(x=Condition,ymin=mean_contamination_gene,ymax=mean_contamination_gene,group=Condition,width=0.5))+
  ylab("% Contaminating Genes")+
  xlab("")+
  scale_colour_manual(values=Contamination_cols)+
  coord_flip()+
  theme(legend.position = "none",
        axis.ticks.x=element_blank())


p.frac.g

```


Barplots species

```{r}
## make list of dfs
path = "/data/share/htp/prime-seq_Paper/Fig_gdna_priming/analysis/count_intergenic/"

files<-list.files(path, recursive = F)

project<-c("rep1","rep2","rep3")

feature_list<- list()

for (i in 1:length(project)){
  feature_list[[i]]<- read.table(paste0(path,project[i],".readspercell.txt"),
                                 sep= "\t",
                                 header=T)
  feature_list[[i]]$XC2<-paste0(feature_list[[i]]$RG,"_",strsplit(project[i],split = "rep")[[1]][2]) ## to make barcodes unique
  names(feature_list)[i]<-project[i]
}


feature_df<-do.call("rbind", feature_list)

feature_df<-left_join(feature_df,sample_info)

feature_df$Replicate<-str_split_fixed(feature_df$XC2,pattern = "_",n=2)[,2]

feature_df2<-feature_df %>% 
  filter(type%in%c("Intergenic_human","Intergenic_mouse"),
         RG!="bad") %>% 
  dplyr::select(XC2,Condition,DNA,N,type) %>% 
  mutate(Type=case_when(type=="Intergenic_human"~"Intergenic:Human",
                        type=="Intergenic_mouse"~"Intergenic:Mouse"),
         nRead=N) %>% 
  dplyr::select(XC2,Condition,DNA,nRead,Type)

### Read counts

exon_list<- list()
for (i in 1:length(dge_list)){
  exon_list[[i]]<- as.matrix(dge_list[[i]]$readcount$exon$all)
  colnames(exon_list[[i]])<-paste0(colnames(exon_list[[i]]),"_",strsplit(project[i],split = "rep")[[1]][2])
  names(exon_list)[i]<-names(dge_list)[i]
}

intron_list<- list()
for (i in 1:length(dge_list)){
  intron_list[[i]]<- as.matrix(dge_list[[i]]$readcount$intron$all)
  colnames(intron_list[[i]])<-paste0(colnames(exon_list[[i]]),"_",strsplit(project[i],split = "rep")[[1]][2])
  names(intron_list)[i]<-names(dge_list)[i]
}

## human only
ENSG_list_ex<- list()
for (i in 1:length(exon_list)){
  ENSG_list_ex[[i]]<- exon_list[[i]][grep("ENSG",row.names(exon_list[[i]])),]
  names(ENSG_list_ex)[i]<-names(exon_list)[i]
}

ENSG_list_in<- list()
for (i in 1:length(intron_list)){
  ENSG_list_in[[i]]<- intron_list[[i]][grep("ENSG",row.names(intron_list[[i]])),]
  names(ENSG_list_in)[i]<-names(intron_list)[i]
}

## mouse only
ENSMUS_list_ex<- list()
for (i in 1:length(exon_list)){
  ENSMUS_list_ex[[i]]<- exon_list[[i]][grep("ENSMUS",row.names(exon_list[[i]])),]
  names(ENSMUS_list_ex)[i]<-names(exon_list)[i]
}

ENSMUS_list_in<- list()
for (i in 1:length(intron_list)){
  ENSMUS_list_in[[i]]<- intron_list[[i]][grep("ENSMUS",row.names(intron_list[[i]])),]
  names(ENSMUS_list_in)[i]<-names(intron_list)[i]
}


df_list<-list()

for (i in 1:length(ENSG_list_ex)){
  df_list[[i]]<- data.frame(XC2=colnames(ENSG_list_ex[[i]]),
                            nRead_HS_ex=colSums(ENSG_list_ex[[i]]),
                            nRead_MM_ex=colSums(ENSMUS_list_ex[[i]]),
                            nRead_HS_in=colSums(ENSG_list_in[[i]]),
                            nRead_MM_in=colSums(ENSMUS_list_in[[i]]))
}

df<- dplyr::bind_rows(df_list)


df2<- dplyr::left_join(df, sample_info)


  

df3<-df2 %>% 
  gather(key = "Type",value="nRead",c(2,3,4,5)) %>% 
  mutate(Type=case_when(
      Type=="nRead_HS_ex"~"Exon:Human",
      Type=="nRead_HS_in"~"Intron:Human",
      Type=="nRead_MM_ex"~"Exon:Mouse",
      Type=="nRead_MM_in"~"Intron:Mouse"),
      nRead=as.integer(nRead)) %>% 
  dplyr::select(XC2,Condition,DNA,nRead,Type) %>% 
  bind_rows(feature_df2) %>% 
  mutate(Type=factor(Type,
                     levels = c("Intergenic:Mouse","Intron:Mouse","Exon:Mouse","Intergenic:Human","Intron:Human","Exon:Human"))) 


p.nreads_inex<-ggplot()+
  geom_bar(data=df3,aes(x=Condition,y=nRead,fill=Type),stat="identity",position=position_fill())+
  coord_flip()+
  scale_fill_manual(values=mouseman_cols)+
  ylab("Assigned Reads")+
  xlab("")+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.text = element_text(size=8),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank())
p.nreads_inex


df3 %>% 
  group_by(XC2) %>% 
  mutate(percent=nRead/sum(nRead)*100,
         total=sum(nRead)) %>% 
ggplot(aes(x=Condition,y=percent,colour=Condition))+
  geom_beeswarm(dodge.width = 0.5)+
  scale_colour_manual(values=Contamination_cols)+
  facet_wrap(~Type,scales = "free")+
  xlab("")+
  ylab("percent of mapped reads")+
  theme(legend.position = "bottom", 
        legend.title = element_blank(),
        legend.text = element_text(size=8),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

```

Phongs DNAse test

```{r}
dnase<-read.table("/data/share/lab/Picogreens/20180816_bulk_DNase_test_phong.txt",sep="\t",header=T)

p.dnase<-dnase %>% 
  mutate(norm_input=((conc*10)/input*1000),
         type=case_when(type=="Dnase"~ "RNA (DNAse treated)",
                        type=="Rnase"~ "gDNA (RNAse treated)",
                        type=="None"~ "RNA + gDNA (untreated)")) %>% 

  ggplot(aes(x=factor(input),y=norm_input))+
  geom_beeswarm(aes(colour=type),dodge.width = 0.5)+
  stat_summary(aes(col=type),fun = mean,geom="crossbar",width=0.3,position=position_dodge(width = 0.5),show.legend = F)+
  xlab("Number of Cells")+
  ylab("Normalized yield\n (ng cDNA per 1000 cells)")+
  scale_color_manual(values=c("#a22edc","#2edca2", "#dca22e"))+
  theme(legend.position = "right", legend.title = element_blank(),legend.text = element_text(size=10))

p.dnase

```


Suplementary Paper figure

```{r}


ggsave(p.dnase,filename = "DNAse_concentration_test.pdf",path = fig_path,device = "pdf",units = "mm",width = 200,height = 100)

p.map2<-p.map+theme(legend.position = "bottom",plot.margin = margin(0.2, 0.3, 2.4, 0, "cm"))+
  guides(fill=guide_legend(direction = "horizontal",reverse = T,nrow=3))

p.nreads_inex2<-p.nreads_inex+theme(plot.margin = margin(0.2, 0.35, 2.4, 0, "cm"))+
  guides(fill=guide_legend(direction = "horizontal",reverse = T,nrow=3))

p.frac2<-p.frac+theme(plot.margin = margin(0.2, 0.3, 5, 0, "cm"))

lower<-plot_grid(p.map2,p.nreads_inex2,p.frac2,ncol=3,rel_widths = c(0.5,0.27,0.25),labels=c("C","D","E"))


ggsave(lower,filename = "Suppl_fig_gdna_lower.pdf",path = fig_path,device = "pdf",units = "mm",width = 300,height = 135)

#individual Figures
ggsave(p.map2,filename = "Assigned_features.pdf",path = fig_path,device = "pdf",units = "mm",width = 150,height = 70)
ggsave(p.nreads_inex,filename = "Mapped_features.pdf",path = fig_path,device = "pdf",units = "mm",width = 100,height = 80)
ggsave(p.frac,filename = "Contaminating_UMIs.pdf",path = fig_path,device = "pdf",units = "mm",width = 100,height = 60)


```

